storage/client_range_test.go

```diff
@@ -150,3 +155,179 @@ func TestRejectFutureCommand(t *testing.T) {
 		t.Errorf("expected 15, got %v", v)
 	}
 }
+
+// TestTxnPutOutOfOrder tests a case where a put operation of an older
+// timestamp comes after a put operation of a newer timestamp in a txn.
+//
+// The test uses "Writer" and "Reader" to reproduce an out-of-order
+// put. Writer executes a put operation in a txn and commits
+// it. Reader executes two get operations on the same key with higher
+// priority (outside of a txn). An out-of-order put when Writer's txn
+// is restarted in the following way:
+//
+// 1) Writer executes a put operation with time T in a txn.
+// 2) Before the writer's txn is committed, Reader sends a get
+//    operation with time T+100. This triggers the restart of
+//    Writer's txn.
+// 3) Before Writer starts a new epoch of the txn, Reader sends
+//    another get operation with time T+200. The write intent is resolved
+//    (and the txn timestamp is pushed again to T+200). The actual get
+//    operation has not yet been executed (and hence the timestamp cache
+//    has not been updated).
+// 4) Writer restarts its txn and executes the put operation
+//    again. This put operation becomes out-of-order since its
+//    timestamp is T+100 while the txn timestamp pushed at Step 3 is T+200.
+func TestTxnPutOutOfOrder(t *testing.T) {
+	defer leaktest.AfterTest(t)
+	manualClock := hlc.NewManualClock(0)
+	clock := hlc.NewClock(manualClock.UnixNano)
+	store, stopper := createTestStoreWithEngine(t,
+		engine.NewInMem(proto.Attributes{}, 10<<20),
+		clock,
+		true,
+		nil)
+	defer stopper.Stop()
+
+	key := "key"
+
+	// Set up a filter to so that the get operation at Step 3 will return an error (simulating
+	// artificial delay).
+	var numGets atomic.Value
```

Comment:
nit: this is probably a better fit for the integer-specific methods in atomic like http://godoc.org/sync/atomic#StoreUint32 and co